src/node_file.cc

```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must not include encoded / characters"));
+
+    return false;
+  }
+
+  *result_file_path = ada::unicode::percent_decode(pathname, pathname.size());
+
+  return true;
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env, const std::string& file_path) {
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemRead, file_path, FilePathIsFileReturnType::kThrowInsufficientPermissions);
+
+  uv_fs_t req;
+
+  int rc = uv_fs_stat(env->event_loop(), &req, file_path.c_str(), nullptr);
+
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    rc = !!(s->st_mode & S_IFDIR);
+  }
+
+  uv_fs_req_cleanup(&req);
+
+  // rc is 0 if the path refers to a file
+  if (rc == 0)
+    return FilePathIsFileReturnType::kIsFile;
+
+  return FilePathIsFileReturnType::kIsNotFile;
+}
+
+static void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
+  // /home/user/project/package.json
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string_view base_path =
+      Utf8Value(env->isolate(), args[0].As<String>()).ToStringView();
+  auto base = ada::parse<ada::url_aggregator>(base_path);
+
+  if (!base) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  ada::url_aggregator* base_pointer = &base.value();
+  ada::url_aggregator* file_path_agg;
+  ada::result<ada::url_aggregator> file_path_url;
+  std::unique_ptr<std::string> initial_file_path;
+  std::string file_path;
+
+  // packageConfig.main
+  // ./index.js
+  // ./lib/index.js
+  if (args[1]->IsString()) {
+    std::string package_config_main =
+        Utf8Value(env->isolate(), args[1].As<String>()).ToString();
+
+    file_path_url = ada::parse<ada::url_aggregator>(
+        std::string("./") + package_config_main, base_pointer);
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+    }
+
+    file_path_agg = &file_path_url.value();
+
+    if (!FileURLToPathImpl(env, file_path_agg, initial_file_path)) return;
+
+    FromNamespacedPath(&(*initial_file_path));
+
+    for (int i = 0; i < legacy_main_extensions_with_main_end; i++) {
+      file_path = *initial_file_path + legacy_main_extensions[i];
+
+      switch(FilePathIsFile(env, file_path)) {
+        case FilePathIsFileReturnType::kIsFile:
+          return args.GetReturnValue().Set(i);
+        case FilePathIsFileReturnType::kIsNotFile:
+          continue;
+        case FilePathIsFileReturnType::kThrowInsufficientPermissions:
+          // the default behavior when do not have permission is to return
+          // and exit the execution of the method as soon as possible
+          // the internal function will throw the exception
+          return;
```

Comment:
It should never execute in this line. I don't think we need this enum.

Reply:
`FilePathIsFile` now verifies the permission model, I need this case because the user can try to import a library using this old method and that library was protected by the permission model, then it should throw an error.